README.md
Description
===========

This is SecondMarket's PostgreSQL "library" cookbook, supporting PostgreSQL 9.0.

Use this cookbook on SecondMarket "platform" servers only. For generic PostgreSQL servers, use the Opscode "postgresql" cookbook instead. The two are mutually exclusive.

Special Note
============

This cookbook is being released publicly as an illustration of how to create "application" cookbooks to overlay community cookbooks as "library cookbooks".

Internal SecondMarket ACLs, feature flags, etc. have been removed for security reasons.

What's a Library Cookbook and What's An Application Cookbook?
=============================================================

Read [Bryan Berry's blog post](http://devopsanywhere.blogspot.com/2012/11/how-to-write-reusable-chef-cookbooks.html) to get an overview.

SecondMarket's platform uses PostgreSQL 9.x, but the underlying operating systems (CentOS 5, CentOS 6) generally ship with PostgreSQL 8.x. Additionally, on "platform" servers we need to set up ACLs, PostgreSQL settings for master/slave replication, etc. and other miscellaneous tuning flags.

Platform
========

Supported only on CentOS 5 and 6, which we use.

Usage
=====

Including `smpostgresql::server` gets you most of what you need.

Credits
=======

* Author:: Julian Dunn (<jdunn@secondmarket.com>)
* Author:: Chris Ferry (<cferry@secondmarket.com>)      

Copyright:: 2012-2013 SecondMarket Labs, LLC.

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


CHANGELOG.md
## unreleased

- Move fix from SOP-2902 to postgresql.conf recipe where it belongs

## v0.1.8

- [SOP-2902] Set vacuum_defer_cleanup_age to 1000 for server

## v0.1.7

- [SOP-2627] Implement 'contrib' support. Upgrade to library cookbook 2.2.2. Fix a few more tests.

## v0.1.6

- [SOP-1342] slave recovery.conf template had a bug

## v0.1.5

- [SOP-2833] Unbreak local access for real

## v0.1.4

- [SOP-2833] Unbreak local access and move Jenkins access to its own feature flag

## v0.1.3

- [SOP-2809] Unbreak direct database access from Jenkins and its nodes

## v0.1.2

- [SOP-1342] Initial release of overlay cookbook


spec/default_spec.rb
require 'spec_helper'

describe 'smpostgresql::default' do
  before do
    Fauxhai.mock(platform: 'centos', version: '6.3') 
  end
  it 'should do nothing' do
    chef_run = ChefSpec::ChefRunner.new.converge 'smpostgresql::default'
    chef_run.should_not log 'anything'
  end
end


spec/pgfouine_spec.rb
require 'spec_helper'

describe 'smpostgresql::pgfouine' do
  let (:chef_run) { ChefSpec::ChefRunner.new.converge 'smpostgresql::pgfouine' }
  it 'should install pgfouine' do
    chef_run.should install_package 'pgfouine'
  end

end


spec/platform_pg_hba_spec.rb
require 'chefspec'

describe 'smpostgresql::platform_pg_hba' do
  let (:chef_run) { ChefSpec::ChefRunner.new.converge 'smpostgresql::platform_pg_hba' }
  it 'should do something' do
    pending 'Your recipe examples go here.'
  end
end


metadata.rb
name              "smpostgresql"
maintainer        "SecondMarket Labs, based on Opscode cookbook"
maintainer_email  "systems@secondmarket.com"
license           "Apache 2.0"
description       "Overlay cookbook over Opscode postgresql for various things that we like"
long_description  IO.read(File.join(File.dirname(__FILE__), 'README.md'))
version           "0.1.8"
recipe            "smpostgresql", "Empty, use one of the other recipes"
recipe            "smpostgresql::client", "Installs postgresql client package(s)"
recipe            "smpostgresql::server", "Installs postgresql server packages, templates"
recipe            "smpostgresql::platform_db_creation", "Create default users and dbs"
recipe            "smpostgresql::pgfouine", "Installs pgfouine query analysis software"
recipe            "smpostgresql::pgbadger", "Installs pgbadger query analysis software"

%w{rhel centos}.each do |os|
  supports os
end

depends           "postgresql", "~> 2.2.2"


recipes/pgdg.rb
#
# Cookbook Name:: smpostgresql
# Recipe:: pgdg
#
# Copyright 2012-2013 SecondMarket Labs, LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

node.set['postgresql']['enable_pgdg_yum'] = true
node.set['postgresql']['version'] = "9.0"
node.set['postgresql']['dir'] = "/var/lib/pgsql/#{node['postgresql']['version']}/data"
node.set['postgresql']['client']['packages'] = ["postgresql#{node['postgresql']['version'].split('.').join}-devel"]
node.set['postgresql']['server']['packages'] = ["postgresql#{node['postgresql']['version'].split('.').join}-server"]
node.set['postgresql']['contrib']['packages'] = ["postgresql#{node['postgresql']['version'].split('.').join}-contrib"]
node.set['postgresql']['server']['service_name'] = "postgresql-#{node['postgresql']['version']}"

include_recipe "postgresql::yum_pgdg_postgresql"


recipes/server.rb
#
# Cookbook Name:: smpostgresql
# Recipe:: server
#
# Copyright 2012-2013, SecondMarket Labs, LLC.
# Copyright 2009-2010, Opscode, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

# Ordering here is very important.
# 1. pgdg is to set up the attributes to overlay Opscode's postgresql cookbook
#    with the version of PostgreSQL we want
# 2. then invoke the platform_postgresql and platform_pg_hba recipes to
#    override Opscode attributes with results from our searches
# 3. once all attribs are set, then call Opscode's recipe

include_recipe "smpostgresql::pgdg"
include_recipe "smpostgresql::platform_pg_hba"
include_recipe "smpostgresql::platform_postgresql_conf"
include_recipe "postgresql::server"

# Jeff Levitt uses some functions imported from tablefunc.sql so it's important to have this
# on the system
include_recipe "postgresql::contrib"

# For WAL archives. Maybe we can turn it off eventually if we parameterize WAL archiving
directory "/var/lib/pgsql/#{node['postgresql']['version']}/archive" do
   action :create
   owner "postgres"
   group "postgres"
   mode 00700
end

# Logrotate setup - consider invoking logrotate cookbook and LWRP down the road
template "/etc/logrotate.d/postgresql" do
  source "postgresql.logrotate.erb"
  mode 00644
  owner "root"
  group "root"
end


recipes/platform_pg_hba.rb
#
# Cookbook Name:: smpostgresql
# Recipe:: platform_pg_hba
#
# Copyright 2012-2013, SecondMarket Labs, LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
#
# Set up ACLs (pg_hba) for SecondMarket platform database servers.

# This is similar to the default pg_hba array that postgresql::server uses, except
# that we use 'md5' instead of 'ident' in the second member (very important)
pg_hba = [
  {:type => 'local', :db => 'all', :user => 'postgres', :addr => nil, :method => 'ident'},
  {:type => 'local', :db => 'all', :user => 'all', :addr => nil, :method => 'md5'},
  {:type => 'host', :db => 'all', :user => 'all', :addr => '127.0.0.1/32', :method => 'md5'},
  {:type => 'host', :db => 'all', :user => 'all', :addr => '::1/128', :method => 'md5'}
]

if Chef::Config['solo']
  Chef::Log.warn("This recipe requires search; Chef Solo doesn't support search.")
  return
end

slave_members = search(:node, "role:postgresql-slave AND chef_environment:#{node.chef_environment}")

slave_members.each do |member|
  pg_hba << {:type => 'host', :db => 'replication', :user => 'postgres', :addr => member['ipaddress'] + '/32', :method => 'trust'}
end

# Rest of ACL recipe elided due to security reasons for publishing on Github

# Do not delete the following for obvious reasons!
node.set['postgresql']['pg_hba'] = pg_hba


recipes/platform_db_creation.rb
#
# Cookbook Name:: smpostgresql
# Recipe:: platform_db_creation
#
# Copyright 2012-2013 SecondMarket Labs, LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

#On initial postgres install this recipe fails as postgres just initdb'd and is not up and listening yet.
# So we need to wait a bit until it comes online
ruby_block "netstat" do
  block do
    10.times do
      if IO.popen("netstat -lnt").entries.select { |entry|
          entry.split[3] =~ /:5432$/
        }.size > 0
        break
      end
      Chef::Log.debug("service[postgresql] still not listening (port 5432)")
      sleep 1
    end
  end
  action :create
end

# mon_user creation
mon_user_password = Chef::EncryptedDataBagItem.load("smpostgresql", "mon_user")['password']

execute "create-mon-database-user" do
    user "postgres"
    command "psql -U postgres -t -c \"CREATE ROLE mon_user PASSWORD '#{mon_user_password}' NOSUPERUSER NOCREATEDB NOCREATEROLE INHERIT LOGIN;GRANT USAGE ON SCHEMA public TO mon_user\""
    not_if 'psql -U postgres -c "SELECT * FROM pg_user WHERE usename=\'mon_user\'" | grep mon_user', :user => "postgres"
end

plat_dbname   = "OBFUSCATED-FOR-GITHUB-PUBLISHING"
integration_dbname   = "OBFUSCATED-FOR-GITHUB-PUBLISHING"

# By default, the usernames for each database are the same as the DB, except in production and staging
plat_password = Chef::EncryptedDataBagItem.load("smpostgresql", plat_dbname)['password']
integration_password = Chef::EncryptedDataBagItem.load("smpostgresql", integration_dbname)['password']

# Plat
execute "create-plat-database-user" do
    user "postgres"
    command "psql -U postgres -t -c \"CREATE ROLE #{plat_dbname} PASSWORD '#{plat_password}' NOSUPERUSER NOCREATEDB NOCREATEROLE INHERIT LOGIN;GRANT USAGE ON SCHEMA public TO #{plat_dbname}\""
    not_if  "psql -U postgres -c \"SELECT * FROM pg_user WHERE usename='#{plat_dbname}'\" | grep #{plat_dbname}", :user => "postgres"
end

execute "create-plat-database" do
    user "postgres"
    command "createdb -U postgres -O #{plat_dbname} -E utf8 -T template0 #{plat_dbname}"
    not_if  "psql -U postgres -c \"SELECT * FROM pg_database WHERE datname='#{plat_dbname}'\" | grep #{plat_dbname}", :user => "postgres"
end
    
# Integration
execute "create-integration-database-user" do
    user "postgres"
    command "psql -U postgres -t -c \"CREATE ROLE #{integration_dbname} password '#{integration_password}' NOSUPERUSER NOCREATEDB NOCREATEROLE INHERIT LOGIN; GRANT USAGE ON SCHEMA public TO #{integration_dbname}\""
    not_if  "psql -U postgres -c \"SELECT * from pg_user WHERE usename='#{integration_dbname}'\" | grep #{integration_dbname}", :user => "postgres"
end

execute "create-integration-database" do
    user "postgres"
    command "createdb -U postgres -O #{integration_dbname} -E utf8 -T template0 #{integration_dbname}"
    not_if  "psql -U postgres -c \"SELECT * from pg_database WHERE datname='#{integration_dbname}'\" | grep #{integration_dbname}", :user => "postgres"
end


recipes/platform_postgresql_conf.rb
#
# Cookbook Name:: smpostgresql
# Recipe:: platform_postgresql_conf
#
# Copyright 2012-2013, SecondMarket Labs, LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
#
# Set up settings that we want in postgresql.conf on platform database servers.

node.set['postgresql']['config']['listen_addresses'] = '*'
node.set['postgresql']['config']['max_connections'] = '256'

# override this to 'minimal' in situations where you don't need full
# WAL archiving, e.g. the slave
node.set['postgresql']['config']['wal_level'] = 'hot_standby'
unless node['postgresql']['config']['wal_level'] == 'minimal'
  node.set['postgresql']['config']['archive_mode'] = true
  node.set['postgresql']['config']['max_wal_senders'] = '5'
  node.set['postgresql']['config']['wal_keep_segments'] = '32'
end

node.set['postgresql']['config']['archive_command'] = "cp %p /var/lib/pgsql/#{node['postgresql']['version']}/archive/%f"

# logging - we like to log to syslog
node.set['postgresql']['config']['log_destination'] = 'syslog'
node.set['postgresql']['config']['syslog_facility'] = 'LOCAL0'
node.set['postgresql']['config']['syslog_ident'] = 'postgres'
node.set['postgresql']['config']['client_min_messages'] = 'notice'
node.set['postgresql']['config']['log_min_messages'] = 'warning'
node.set['postgresql']['config']['log_min_duration_statement'] = '100'
node.set['postgresql']['config']['log_line_prefix'] = '%t %d %u %h '
# node.set['postgresql']['config']['track_counts'] = 'on' # already on by default
node.set['postgresql']['config']['autovacuum'] = true
node.set['postgresql']['config']['log_autovacuum_min_duration'] = '-1'

# SOP-2902: PostgreSQL replication isn't great, and any rows removed from the master by the autovacuum process
# may cause long-running slave queries to fail, so defer autovacuum cleanup by 1000 transactions.
node.set['postgresql']['config']['vacuum_defer_cleanup_age'] = '1000'
